lib/evaluate/character.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_CHARACTER_H_
+#define FORTRAN_EVALUATE_CHARACTER_H_
+
+#include "type.h"
+#include <string>
+
+// character(1) is ISO/IEC 646:1991 (ASCII)
+//
+// character(2) is a variant of EUC-JP: The code points are the same as EUC-JP,
+// but the encoding has a two bytes fix-size (EUC-JP encoding has a variable
+// length). The one byte EUC-JP character representations are simply
+// zero-extended to two byte representations. The three bytes character
+// representation of EUC-JP (JIS X 0212) are not supported in this internal
+// encoding.
+//
+// character(4) is ISO/IEC 10646 UCS-4 (~ UTF-32)
+
+namespace Fortran::evaluate {
+
+template<int KIND> class CharacterUtils {
+  using Character = Scalar<Type<TypeCategory::Character, KIND>>;
+  using CharT = typename Character::value_type;
+
+public:
+  static constexpr bool IsValidCharacterCode(std::uint64_t code) {
+    if constexpr (KIND == 1) {
```

Comment:
Do we want to hard-code these kind values or move them to a config file?  Is the const-ness of these functions important?

Reply:
Regarding the constexpr-ness, it is not required by call-site, I did put it because I could and this follows our style guideline to do so when possible.
So, yes, I think it is possible here to move the mapping between the kind and the encoding into some kind of config file and have the semantic or folding context initialized with the mapping.

However, I think the semantic analysis phase is already making some assumptions regarding the encoding behind the character kinds. Also, making the size of the encoding configurable would be troublesome (1 is one byte, 2 is two byte and 4 is four). So people could only configure 2 to be UCS-2 or the EUC-JP variant or another 2-byte encoding.
 
By config file, do you mean a config file to be read by f18 before every compilation of a Fortran program or a config file that is used once during the compilation of f18 ?

Reply:
By config file, I just mean that all implementation-dependent constants ought to be collected in one place.  In a C++ source file is fine with me.

Reply:
Then `lib/evaluate/type.h` might be the good place to map the character KIND to encodings, since it already gathers this kind of choice for numerical types. Does that sounds right to you ?

Reply:
OK w/ me.

Reply:
I tracked back from moving the KIND mapping in `lib/evaluate/type.h` following comments from Tim and Peter below. I removed checks for encoding validity so there is no need for any kind of mapping here anymore.